File: 2023/day12/part1.py
from itertools import product

from tqdm import tqdm

with open("./input.txt", "r") as file:
    lines = list(map(lambda s: s.removesuffix("\n"), file.readlines()))


def compute_size_contiguous_groups(line: str):
    res: list[int] = []
    current = 0
    for char in line:
        if char == "#":
            current += 1
        elif current != 0:
            res.append(current)
            current = 0
    if current != 0:
        res.append(current)
    return res


total = 0

for line in tqdm(lines):
    text, nbrs = line.split(" ")
    nbrs = list(map(int, nbrs.split(",")))
    unknowns: list[int] = [i for i, char in enumerate(text) if char == "?"]
    for combination in product([True, False], repeat=len(unknowns)):
        true_unknown = [n for i, n in enumerate(unknowns) if combination[i]]
        new_text = "".join(
            ["#" if i in true_unknown else char for i, char in enumerate(text)]
        )
        new_nbrs = compute_size_contiguous_groups(new_text)
        if all(
            len(new_nbrs) == len(nbrs) and new_nbrs[i] == nbr
            for i, nbr in enumerate(nbrs)
        ):
            total += 1

print(total)

File: 2021/day3/part1.ts
import * as fs from "fs/promises"

const input = await fs.readFile("input.txt", "utf8")
const lines = input.split("\n")

const mostCommonBits = new Array(lines[0].length).fill(0)
lines.forEach((line) => {
  line.split("").forEach((char, i) => {
    mostCommonBits[i] += char === "1" ? 1 : -1
  })
})

const gammaRate = parseInt(mostCommonBits.map((n) => (n > 0 ? 1 : 0)).join(""), 2)
const epsilonRate = parseInt(mostCommonBits.map((n) => (n < 0 ? 1 : 0)).join(""), 2)
console.log(gammaRate)
console.log(epsilonRate)
console.log(gammaRate * epsilonRate)
